luaui/Widgets/unit_dgun_stall_assist.lua

```diff
@@ -1,59 +1,88 @@
+end
 
-----------------------------------------------------------------
--- Globals
-----------------------------------------------------------------
-local watchTime = 0
-local waitedUnits = nil -- nil / waitedUnits[1..n] = uID
-local shouldWait = {} -- shouldWait[uDefID] = true / nil
-local isFactory = {} -- isFactory[uDefID] = true / nil
+local target_energy = 1000
+local dgun_cost = 500
+local time_in_ms_befor_unwaiting = 250 --how many fast we can unwait units after waiting them
+local gameStarted = false
+local DGUN_OUT = false
+local UNITS_WAITING = false
+local waitableUnits = {}
+local waitingUnits = {}
+
+local function getWaitAbleUnits()
+  local myWaitableUnits = {}
+  local myUnits = Spring.GetTeamUnits(Spring.GetMyTeamID())
+  for i=1, #myUnits do
+    local unitCmd = Spring.GetUnitCurrentCommand(myUnits[i])
+    local isEnergyConsumingCommand = not (unitCmd == CMD.WAIT or unitCmd == CMD.MOVE or unitCmd == CMD.RECLAIM)
+    if isEnergyConsumingCommand then
+      local unitDefId = Spring.GetUnitDefID(myUnits[i])
+      if waitableUnits[unitDefId] then myWaitableUnits[#myWaitableUnits + 1] = myUnits[i] end
+    end
+  end
+  return myWaitableUnits
+end
 
-local gameStarted
+local function doWeHaveLowEnergy()
+  local currentEnergy, energyStorage = Spring.GetTeamResources(Spring.GetMyTeamID(), "energy")
+  if energyStorage < dgun_cost then return false end
+  if currentEnergy < target_energy then return true end
+  return false
+end
+
+local next_call_when = 0
+local function wait_units()
+  local time_now = millis()
+  if next_call_when > time_now then return end
+  next_call_when = time_now + time_in_ms_befor_unwaiting
+  UNITS_WAITING = true
+  for _, unitID in pairs(getWaitAbleUnits()) do
+    waitingUnits[#waitingUnits +1] = unitID
+  end
+  Spring.GiveOrderToUnitArray(waitingUnits, CMD.WAIT, {}, 0)
+end
+
+local function unwait_units()
+  Spring.GiveOrderToUnitArray(waitingUnits, CMD.WAIT, {}, 0)
+  UNITS_WAITING = false
+  waitingUnits = {}
+end
 
-local stallIds = {UnitDefNames['armcom'].id, UnitDefNames['corcom'].id}
+local function buildWaitableUnitsTable()
+  for uDefID, uDef in pairs(UnitDefs) do
+		if (uDef.buildSpeed > 0) and uDef.canAssist and (not uDef.canManualFire) then
+			waitableUnits[uDefID] = true
+		end
+	end
+end
 
-----------------------------------------------------------------
--- Speedups
-----------------------------------------------------------------
-local spGetActiveCommand = Spring.GetActiveCommand
-local spGiveOrderToUnitArray = Spring.GiveOrderToUnitArray
-local spGetUnitCurrentCommand = Spring.GetUnitCurrentCommand
-local spGetFactoryCommands = Spring.GetFactoryCommands
-local spGetMyTeamID = Spring.GetMyTeamID
-local spGetTeamResources = Spring.GetTeamResources
-local spGetTeamUnits = Spring.GetTeamUnits
-local spGetUnitDefID = Spring.GetUnitDefID
-local spGetSpectatingState = Spring.GetSpectatingState
-local spGetUnitIsBeingBuilt = Spring.GetUnitIsBeingBuilt
+function widget:GameFrame()
```

Comment:
We do not want to do this in `GameFrame`, it'll run way too often. Do it in the update loop. The timing here felt very overcomplicated to me, I have simplified it to this:
```lua
function widget:Update()
	timer = timer + Spring.GetLastUpdateSeconds()
	local lowEnergy = doWeHaveLowEnergy()
	local _, activeCmdID = Spring.GetActiveCommand()
	dgunActive = activeCmdID == CMD.DGUN
	if dgunActive then
		lastDgunTime = timer
	end

	if dgunActive and lowEnergy and #waitingUnits == 0 then
		waitUnits()
		return
	end

	if
		not dgunActive and
		#waitingUnits > 0 and
		lastDgunTime > 0 and
		lastDgunTime < timer - unwaitThreshold
	then
		unwaitUnits()
	end
end
```